game-core/src/main/java/org/triplea/performance/PerfTimer.java

```diff
@@ -23,52 +24,88 @@
 @Log
 public class PerfTimer implements Closeable {
 
-  private static final PerfTimer DISABLED_TIMER = new PerfTimer("disabled");
   private static final Map<String, AtomicLong> runningTotal = new HashMap<>();
-  final String title;
-  private final long startMillis;
+  private static final Map<String, AtomicLong> runningCount = new HashMap<>();
+  private final String title;
+  private final Optional<Long> startNanos;
+  private final int reportingFrequency;
 
-  private PerfTimer(final String title) {
+  private PerfTimer(final String title, final int reportingFrequency) {
     this.title = title;
-    this.startMillis = System.nanoTime();
+    this.reportingFrequency = reportingFrequency;
+    // If reporting frequency is 0, avoid the overhead of querying the time.
+    if (this.reportingFrequency > 0) {
+      this.startNanos = Optional.of(System.nanoTime());
+    } else {
+      this.startNanos = Optional.empty();
+    }
   }
 
   private long stopTimer() {
-    return System.nanoTime() - startMillis;
+    return System.nanoTime() - startNanos.get();
   }
 
   @Override
   public void close() {
-    processResult(stopTimer(), this);
+    if (this.reportingFrequency > 0) {
+      processResult(stopTimer(), this);
+    }
   }
 
   @SuppressWarnings("unused")
   public static PerfTimer startTimer(final String title) {
-    return new PerfTimer(title);
+    return startTimer(title, 1);
   }
 
-  private static synchronized void processResult(final long stopNanos, final PerfTimer perfTimer) {
-    final long stopMicros = stopNanos / 1000;
-
-    final long milliFraction = (stopMicros % 1000) / 100;
-    final long millis = (stopMicros / 1000);
+  /**
+   * Creates a perf timer with a reporting frequency. The reporting frequency specifies N specifies
+   * that performance information should be printed every N executions of the timer. If 0, no
+   * information is printed (and no timings are taken), which can be useful to have some places in
+   * the code to be always instrumented, but not always enabled.
+   *
+   * @param title The name of the timer
+   * @param reportingFrequency The reporting frequency.
+   * @return the perf timer object
+   */
+  @SuppressWarnings("unused")
+  public static PerfTimer startTimer(final String title, final int reportingFrequency) {
+    return new PerfTimer(title, reportingFrequency);
+  }
 
+  private static synchronized void processResult(final long stopNanos, final PerfTimer perfTimer) {
```

Comment:
Have you checked the `synchronized` block on highly concurrent code? I had some problems with this for tile rendering and noticed that the overhead of closing the timer with `processResult` slowed things down to the point where the map was taking seconds to render. Just curious what scenario's you've run this through.

I think the concurrency problem here might be just tricky (argument for us to get a library to do this).

Reply:
Agreed that it might be a problem. Note: My PR is not changing this, however - it was that way before.